include/seqan/vcf_io/read_vcf.h

```diff
@@ -64,21 +64,43 @@ typedef Tag<Vcf_> Vcf;
 // Function readRecord()                                            [VcfHeader]
 // ----------------------------------------------------------------------------
 
+template <typename TNameStore, typename TNameStoreCache, typename TStorageSpec, typename TString>
 inline void
-_parseVcfContig(CharString & chromName, CharString const & headerValue)
+_readVcfContig(VcfIOContext<TNameStore, TNameStoreCache, TStorageSpec> & context, TString const & headerValue)
 {
-    if (length(headerValue) < 3u)
-        return;
+    typedef OrFunctor<EqualsChar<','>, EqualsChar<'>'> >            IsCommaOrGt;
+    typedef typename DirectionIterator<TString const, Input>::Type  TIter;
 
-    CharString tmp = infix(headerValue, 1, length(headerValue) - 2);
-    StringSet<CharString> tmp2;
-    strSplit(tmp2, tmp, EqualsChar<','>());
-    for (unsigned i = 0; i < length(tmp2); ++i)
-    {
-        if (!startsWith(tmp2[i], "ID="))
-            continue;
-        chromName = suffix(tmp2[i], 3);
-    }
+    TIter headerIter = directionIterator(headerValue, Input());
+    CharString &buffer = context.buffer;
+
+    // Seek ID key.
+    clear(buffer);
+    skipOne(headerIter);
+    readUntil(buffer, headerIter, EqualsChar<'='>());
+    if (buffer != "ID")
+        SEQAN_THROW(ParseError("Contig ID key not found in header."));
+    skipOne(headerIter);
+
+    // Read ID value.
+    clear(buffer);
+    readUntil(buffer, headerIter, IsCommaOrGt());
+    if (empty(buffer))
+        SEQAN_THROW(ParseError("Contig ID key is empty."));
+    appendName(contigNamesCache(context), buffer);
+
```

Comment:
Mhmm, I think it is not specified, whether ID key is the first or length key is the second element in the contig field, is it? So it could lead to an error, when the order of the tags is mixed. Or is it somewhere explicitly specified? I just could find for vcf4.3 that the ID tag is required. So maybe adapt it to parse all tags and then check for the corresponding key.